module/x/gravity/keeper/grpc_query.go

```diff
@@ -471,3 +507,129 @@ func (k Keeper) LastObservedEthereumHeight(c context.Context, req *types.LastObs
 
 	return res, nil
 }
+
+func (k Keeper) CompletedOutgoingTxs(c context.Context, req *types.CompletedOutgoingTxsRequest) (*types.CompletedOutgoingTxsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	var batches []*types.BatchTx
+	var contractCalls []*types.ContractCallTx
+	var signerSetCalls []*types.SignerSetTx
+	k.IterateCompletedOutgoingTxs(ctx, func(_ []byte, tx types.OutgoingTx) bool {
+		switch tx := tx.(type) {
+		case *types.BatchTx:
+			batches = append(batches, tx)
+		case *types.ContractCallTx:
+			contractCalls = append(contractCalls, tx)
+		case *types.SignerSetTx:
+			signerSetCalls = append(signerSetCalls, tx)
+		}
+		return false
+	})
+
+	res := &types.CompletedOutgoingTxsResponse{
+		CompletedBatchTxs:     batches,
+		CompletedLogicCalls:   contractCalls,
+		CompletedSignerSetTxs: signerSetCalls,
+	}
+	return res, nil
+}
+
+func (k Keeper) BatchTxConfirmationsByValidator(c context.Context, req *types.BatchTxConfirmationsByValidatorRequest) (*types.BatchTxConfirmationsByValidatorResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	valAddr, err := sdk.ValAddressFromBech32(req.ValidatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	var confirms []*types.BatchTxConfirmation
+	k.IterateBatchTxEthereumSignatures(ctx, func(contractAddress common.Address, nonce uint64, val sdk.ValAddress, sig []byte) bool {
```

Comment:
Do you mean like moving part of the internals of these lines into an `IterateBatchTxConfirmations` method?

Reply:
Basically this iterator is going over every `BatchTxConfirmation`, and providing to the `func` argument some of the contents of each element, then we reconstruct those `BatchTxConfirmation` structs from the individual elements. I guess maybe this is unavoidable because we need to inspect `val`, but it just seems a bit cumbersome to deconstruct the entire thing and then recreate it.

Reply:
I don't like it either. It's because of the generalization of the "confirmation" concept I think. Since each type has different fields, we can only save the val/signature in state, not an entire confirmation, without splitting them out explicitly in state. 

```go
func (k Keeper) SetEthereumSignature(ctx sdk.Context, sig types.EthereumTxConfirmation, val sdk.ValAddress) []byte {
	key := types.MakeEthereumSignatureKey(sig.GetStoreIndex(), val)
	ctx.KVStore(k.storeKey).Set(key, sig.GetSignature())
	return key
}
```